File: src/day12/day12.py
from functools import cache


@cache
def calc_ways(is_op_must, is_dam_must, rcrd, grps):
    if rcrd == () and grps == ():
        return 1
    if rcrd == ():
        return 0
    r, *rs = rcrd
    rs = tuple(rs)
    if r == "." and is_dam_must:
        return 0
    if r == ".":
        return calc_ways(False, False, rs, grps)
    if r == "#" and is_op_must:
        return 0
    if r == "#" and not grps:
        return 0
    if r == "#" and grps[0] > 1:
        return calc_ways(False, True, rs, (grps[0] - 1, *grps[1:]))
    if r == "#" and grps[0] == 1:
        return calc_ways(True, False, rs, grps[1:])
    if r == "?" and is_dam_must:
        return calc_ways(is_op_must, is_dam_must, ("#",) + rs, grps)
    if r == "?" and is_op_must:
        return calc_ways(is_op_must, is_dam_must, (".",) + rs, grps)
    if r == "?":
        return calc_ways(is_op_must, is_dam_must, ("#",) + rs, grps) + calc_ways(
            is_op_must, is_dam_must, (".",) + rs, grps
        )


def main():
    with open("input.txt") as f:
        inp = f.read().splitlines()

    parsed2 = []
    for i in inp:
        x, y = i.split()
        x = ((tuple(x) + ("?",)) * 5)[:-1]
        y = tuple(map(int, y.split(","))) * 5
        parsed2.append((x, y))
    print(sum([calc_ways(False, False, r, g) for r, g in parsed2]))


main()

File: src/day12/Day12.hs
import Data.List (intercalate)
import Data.List.Split (splitOn)
import Data.Map (Map)

main :: IO ()
main = do
  contents <- readFile "input.txt"
  print $ part1 contents

parseLine :: String -> (String, [Int])
parseLine ln =
  let [record, grps] = splitOn " " ln
   in (record, map read $ splitOn "," grps)

part1 :: String -> Int
part1 contents = sum $ map (uncurry (calcWays False False) . parseLine) $ lines contents

calcWays :: Bool -> Bool -> String -> [Int] -> Int
calcWays isOpMust isDamMust rcrd@(r : rs) grps
  | r == '.' && isDamMust = 0
  | r == '.' = calcWays False False rs grps
  | r == '#' && isOpMust = 0
  | r == '#' && null grps = 0
  | r == '#' && head grps > 1 = calcWays False True rs (head grps - 1 : tail grps)
  | r == '#' && head grps == 1 = calcWays True False rs (tail grps)
  | r == '?' && isDamMust = calcWays isOpMust isDamMust ('#' : rs) grps
  | r == '?' && isOpMust = calcWays isOpMust isDamMust ('.' : rs) grps
  | r == '?' = calcWays isOpMust isDamMust ('#' : rs) grps + calcWays isOpMust isDamMust ('.' : rs) grps
calcWays _ _ [] [] = 1
calcWays _ _ [] _ = 0

parseLine2 :: String -> (String, [Int])
parseLine2 contents =
  let (rcrd, grps) = parseLine contents
      rcrd' = intercalate "?" (replicate 5 rcrd)
      grps' = concat (replicate 5 grps)
   in (rcrd', grps')
